Update the docs about setting up terraform (#1595)
2i2c-org/infrastructure: docs/topic/terraform.md

- [`terraform/gcp`](https://github.com/2i2c-org/infrastructure/tree/HEAD/terraform/gcp)
- [`terraform/azure`](https://github.com/2i2c-org/infrastructure/tree/HEAD/terraform/azure)
- [`terraform/aws`](https://github.com/2i2c-org/infrastructure/tree/HEAD/terraform/aws)

Each provider directory also has a `projects` subdirectory where variables defining each cluster are stored.

## Backends

> A [backend](https://www.terraform.io/language/settings/backends/configuration) defines where Terraform stores its state data files. Normally, terraform automatically manages state in the workspaces associated with your configuration, and there's no need to explicity define a backend.

Reviewer: ```suggestion
A [backend](https://www.terraform.io/language/settings/backends/configuration) defines where Terraform stores its state data files in cloud storage (we use a GCP bucket for the majority of our deployments).
Normally, we are using Terraform when logged in with our 2i2c accounts and there's no need to explicitly define a backend.
```

But for some projects where we don't have access to using our 2i2c accounts, e.g. universities that require us to have specific university-affiliated identities, we can configure different backends to access the terraform state stored in those projects.
Working this way saves us the pain of trying to work with terraform using two different authentications.

> These backend configs are stored in [`terraform/gcp/backends`](https://github.com/2i2c-org/infrastructure/tree/HEAD/terraform/gcp/backends) 

Reviewer: ```suggestion
These backend configs are stored in [`terraform/gcp/backends`](https://github.com/2i2c-org/infrastructure/tree/HEAD/terraform/gcp/backends).
Currently, we only implement a choice of backend for GCP deployments.
```

### Backend Initialization

To initialize the backends, you need to run `terraform init` command from within the directory where the source files of the specific provider you want to access are.

Example:

```bash
cd terraform/azure
terraform init
```

````{note}
Currently, since our GCP terraform config is the only one that uses different backends, an extra `-backend-config` flag needs to be passed to the `init` command to initialize.

For example, for our Pangeo projects, run:

```bash
cd terraform/gcp
terraform init -backend-config=backends/pangeo-backend.hcl
```
````

```{note}

> If prior backend data exists in a `terraform.lock.hcl`, you might see an `ERROR` when trying to initialize that backend. To reconfigure this backend, ignoring any saved configuration, add the `-reconfigure` flag to the init command.

Reviewer: Can we remember what the error actually was and add is here instead of `ERROR`

Author: I'm not sure I ever seen that error, hmm. I believe I always used the `-reconfigure` flag in my terraform init command, because I just copy pasted it from the docs, without putting any thought into it 😅 

Maybe it's this one I found on the internet (https://stackoverflow.com/questions/68159962/cannot-convert-terraform-remote-state-in-s3-backend-to-local-state-with-terrafo)?:

```
Error: Backend configuration changed
│ 
│ A change in the backend configuration has been detected, which may require migrating existing state.
│ 
│ If you wish to attempt automatic migration of the state, use "terraform init -migrate-state".
│ If you wish to store the current configuration with no changes to the state, use "terraform init -reconfigure".
```

Reviewer: Yes, I think that's the one! I think you could just swap out `ERROR` for `Error: Backend configuration changed` rather than paste the whole thing, as we are providing the information about using the `-reconfigure` flag as the solution

Author: ```suggestion
If prior backend data exists in a `terraform.lock.hcl`, you might see an `Error: Backend configuration changed` when trying to initialize that backend. To reconfigure this backend, ignoring any saved configuration, add the `-reconfigure` flag to the init command.
```
